taichi/analysis/same_statements.cpp

```diff
@@ -39,18 +43,29 @@ class IRNodeComparator : public IRVisitor {
     }
   }
 
-  int get_other_id(int this_id) {
+  void check_mapping(Stmt *this_stmt, Stmt *other_stmt) {
     // get the corresponding id in the other node
-    auto it = id_map.find(this_id);
+    // and check if it is other_stmt->id
+    auto it = id_map.find(this_stmt->id);
     if (it != id_map.end()) {
-      return it->second;
+      if (it->second != other_stmt->id) {
+        same = false;
+      }
+      return;
     }
-    // if not found, should be captured
-    // (What if this_id belongs to the other node? Ignoring this case here.)
-    if (captured_id.find(this_id) == captured_id.end()) {
-      captured_id.insert(this_id);
+    if (implicitly_map_ids) {
```

Comment:
Not sure if this make sense: would `build_mapping_on_the_fly` be a better name compared to `implicitly_map_ids`?

Reply:
If this boolean variable is false, we still build the mapping in `map_id`...

This boolean variable actually does early rejecting -- we will just return "not same" if this boolean variable is true and `root1` and `root2` have different operands.

Reply:
Renamed to `recursively_check_` and added more comments.